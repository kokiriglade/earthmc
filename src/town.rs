//! # Town
//!
//! Defines the [`Town`] struct.
use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};
use uuid::Uuid;

use crate::{
    chunk_position::ChunkPosition,
    named_id::{NamedId, NamedIdOpt},
    permission::TownyPermissions,
    world_location::WorldLocation,
};

#[derive(Deserialize, Serialize, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub struct Town {
    /// The town's name.
    pub name: String,
    /// The town's [`Uuid`].
    pub uuid: Uuid,
    /// The town's board as seen on `/t` in-game.
    pub board: Option<String>,
    /// The founder of the town as seen on `/t` in-game.
    pub founder: String,
    /// The town's wiki URL.
    pub wiki: Option<String>,
    /// The town's mayor.
    pub mayor: NamedId,
    /// The town's nation.
    pub nation: NamedIdOpt,
    pub timestamps: TownTimestamps,
    pub status: TownStatus,
    pub stats: TownStats,
    pub perms: TownyPermissions,
    pub coordinates: TownCoordinates,
    /// A list of all of the residents of the town.
    pub residents: Vec<NamedId>,
    /// A list of all of the trusted residents of the town.
    pub trusted: Vec<NamedId>,
    /// A list of all of the outlawed people of the town.
    pub outlaws: Vec<NamedId>,
    /// A list of every Quarter in the town.
    pub quarters: Vec<NamedId>,
    /// Lists of everyone in every rank in the town.
    pub ranks: BTreeMap<TownRankKind, Vec<NamedId>>,
}

#[derive(Deserialize, Serialize, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub struct TownTimestamps {
    /// Unix timestamp representing when the town was created.
    pub registered: i64,
    /// Unix timestamp representing when the town joined its current nation.
    pub joined_nation_at: i64,
    /// Unix timestamp representing when the town fell into ruin.
    pub ruined_at: Option<i64>,
}

#[derive(Deserialize, Serialize, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub struct TownStatus {
    /// If the town is public.
    pub is_public: bool,
    /// If the town is open (invite-less joining enabled).
    pub is_open: bool,
    /// If the town is neutral.
    pub is_neutral: bool,
    /// If the town is the nation's capital.
    pub is_capital: bool,
    /// If the town ahs more claims than it should.
    pub is_over_claimed: bool,
    /// If the town is ruined.
    pub is_ruined: bool,
    /// If the town is for sale.
    pub is_for_sale: bool,
    /// If the town has a nation.
    pub has_nation: bool,
    /// If the town currently has an overclaim shield.
    pub has_overclaim_shield: bool,
    /// If the town allows outsiders to teleport to its spawn point.
    pub can_outsiders_spawn: bool,
}

#[derive(Deserialize, Serialize, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub struct TownStats {
    /// The total number of town blocks the town has.
    pub num_town_blocks: i32,
    /// The maximum town blocks the town can claim.
    pub max_town_blocks: i32,
    /// The current number of residents in the town.
    pub num_residents: i32,
    /// The total number of trusted residents in the town.
    pub num_trusted: i32,
    /// The total number of players that are outlawed in the town.
    pub num_outlaws: i32,
    /// The town's balance as seen on `/t` in-game.
    pub balance: f64,
    /// The price the town is for sale at if it is for sale.
    pub for_sale_price: Option<f64>,
}

#[derive(Deserialize, Serialize, PartialEq, Debug)]
#[serde(rename_all = "camelCase")]
pub struct TownCoordinates {
    /// The locatioin of the town's spawn point.
    pub spawn: WorldLocation,
    pub home_block: ChunkPosition,
    pub town_blocks: Vec<ChunkPosition>,
}

#[derive(
    Deserialize, Serialize, PartialEq, Eq, Hash, PartialOrd, Ord, Debug,
)]
#[serde(rename_all = "PascalCase")]
pub enum TownRankKind {
    Councillor,
    Builder,
    Recruiter,
    Police,
    #[serde(rename = "Tax-exempt")]
    TaxExempt,
    Treasurer,
    Realtor,
    Settler,
}
