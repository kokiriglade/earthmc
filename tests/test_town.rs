use earthmc::town::Town;

#[test]
fn test_town() {
    let raw_json = include_str!("inputs/town.json");
    let parsed: Vec<Town> = serde_json::from_str(raw_json).unwrap();
    insta::assert_debug_snapshot!(parsed);
}
