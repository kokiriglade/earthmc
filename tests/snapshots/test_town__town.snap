---
source: tests/test_town.rs
expression: parsed
---
[
    Town {
        name: "Teslin",
        uuid: 969d4960-f082-4be0-9e32-f14ed0b7cd50,
        board: Some(
            "Fishing every Friday! Join our Discord, link can be found on the signs!",
        ),
        founder: "tuzzzie",
        wiki: None,
        mayor: NamedIdGeneric {
            name: "Fruitloopins",
            uuid: fed0ec4a-f1ad-4b97-9443-876391668b34,
        },
        nation: NamedIdGeneric {
            name: Some(
                "Yukon",
            ),
            uuid: Some(
                3ce63924-44f9-404a-93e0-2fe87a22edd9,
            ),
        },
        timestamps: TownTimestamps {
            registered: 1660454740242,
            joined_nation_at: 1660454906803,
            ruined_at: None,
        },
        status: TownStatus {
            is_public: false,
            is_open: true,
            is_neutral: false,
            is_capital: true,
            is_over_claimed: true,
            is_ruined: false,
            is_for_sale: false,
            has_nation: true,
            has_overclaim_shield: false,
            can_outsiders_spawn: false,
        },
        stats: TownStats {
            num_town_blocks: 473,
            max_town_blocks: 114,
            num_residents: 7,
            num_trusted: 5,
            num_outlaws: 3,
            balance: 51.0,
            for_sale_price: None,
        },
        perms: TownyPermissions {
            build: TownyPermissionSet {
                resident: false,
                nation: false,
                ally: false,
                outsider: false,
            },
            destroy: TownyPermissionSet {
                resident: false,
                nation: false,
                ally: false,
                outsider: false,
            },
            switch: TownyPermissionSet {
                resident: false,
                nation: false,
                ally: false,
                outsider: false,
            },
            item_use: TownyPermissionSet {
                resident: false,
                nation: false,
                ally: false,
                outsider: false,
            },
            flags: TownyPermissionFlags {
                pvp: false,
                explosion: false,
                fire: false,
                mobs: false,
            },
        },
        coordinates: TownCoordinates {
            spawn: Location {
                world: "earth",
                x: -25227.148,
                y: 86.0,
                z: -11394.415,
                pitch: 16.419712,
                yaw: -60.628418,
            },
            home_block: ChunkPosition {
                x: -1577,
                y: -713,
            },
            town_blocks: [
                ChunkPosition {
                    x: -1551,
                    y: -710,
                },
                ChunkPosition {
                    x: -1551,
                    y: -709,
                },
                ChunkPosition {
                    x: -1551,
                    y: -712,
                },
                ChunkPosition {
                    x: -1551,
                    y: -711,
                },
                ChunkPosition {
                    x: -1551,
                    y: -714,
                },
            ],
        },
        residents: [
            NamedIdGeneric {
                name: "SilentlyyJudging",
                uuid: 00abc813-22c4-4734-88a5-609821971147,
            },
            NamedIdGeneric {
                name: "Fruitloopins",
                uuid: fed0ec4a-f1ad-4b97-9443-876391668b34,
            },
            NamedIdGeneric {
                name: "catgirljodie",
                uuid: 6ff64e22-8d20-40bf-9914-a2d6172e65cc,
            },
            NamedIdGeneric {
                name: "AuroraServer",
                uuid: d633b540-2509-4aeb-af94-ea9ea2c481ee,
            },
            NamedIdGeneric {
                name: "tuzzzie",
                uuid: 8391474f-4b57-412a-a835-96bd2c253219,
            },
            NamedIdGeneric {
                name: "Thaamoth",
                uuid: fa61c5dc-2677-4509-b20e-d159da20c9fd,
            },
            NamedIdGeneric {
                name: "rangegirl28",
                uuid: 6ae5516a-c5e6-488f-97a4-311aaf6ecced,
            },
        ],
        trusted: [
            NamedIdGeneric {
                name: "Fruitloopins",
                uuid: fed0ec4a-f1ad-4b97-9443-876391668b34,
            },
            NamedIdGeneric {
                name: "Masrain",
                uuid: ea0c4d98-106d-416c-9647-06203f861e1b,
            },
            NamedIdGeneric {
                name: "CorruptedGreed",
                uuid: f17d77ab-aed4-44e7-96ef-ec9cd473eda3,
            },
            NamedIdGeneric {
                name: "HaisevaGreed",
                uuid: 253dd297-cb8d-4b9a-9aae-ffe71883956f,
            },
            NamedIdGeneric {
                name: "tuzzzie",
                uuid: 8391474f-4b57-412a-a835-96bd2c253219,
            },
        ],
        outlaws: [
            NamedIdGeneric {
                name: "Fruitloopins",
                uuid: fed0ec4a-f1ad-4b97-9443-876391668b34,
            },
            NamedIdGeneric {
                name: "Masrain",
                uuid: ea0c4d98-106d-416c-9647-06203f861e1b,
            },
            NamedIdGeneric {
                name: "CorruptedGreed",
                uuid: f17d77ab-aed4-44e7-96ef-ec9cd473eda3,
            },
            NamedIdGeneric {
                name: "HaisevaGreed",
                uuid: 253dd297-cb8d-4b9a-9aae-ffe71883956f,
            },
            NamedIdGeneric {
                name: "tuzzzie",
                uuid: 8391474f-4b57-412a-a835-96bd2c253219,
            },
        ],
        quarters: [
            8a7c56d1-d45a-4f5f-ac42-751c2e8c1b3d,
            aeac5e83-62e5-421d-9737-9f33825479ec,
            49a4da60-4589-45ff-a0ba-0fd54c45ba9f,
            54177617-694d-43b1-b2d9-2bb553fd8775,
            fc8a7c21-2992-4584-970c-e0b98c2233cd,
            47cf0463-dd50-4b2b-b533-5b57b3a51c09,
            853908a4-2d76-4055-b83a-cee9cda714ce,
            95d00017-845f-4ccd-883a-9104aabc1391,
            473c2da9-e0b1-467a-87c8-98a59b62e386,
            7645cc85-0d54-4ce0-800e-d358ae647aac,
            0253c695-15aa-48a6-9592-d558224dcff7,
            66ec664a-7eb9-47a6-aa53-1b692ef37e5e,
            3c130442-4521-47f2-9eb9-2f61db1c3b58,
            7fc3351c-42b0-4d55-b309-5ea7325519d4,
            0624ab20-19c8-4277-a900-5f336417ad67,
            93bd12da-9854-41eb-ad35-fb7789e47811,
            c5190a18-2001-475a-a9cc-b584998e6d91,
            9d966cc2-61c4-45e6-a792-45cd6a5affb6,
            fed851f7-09f7-49c4-9a12-51b22834145f,
            e8e594bd-2651-4559-a717-dbd75c125de6,
            439b01d5-a2b7-4a3c-8cb5-3e8389e1215e,
            f244d142-a0e0-4bf5-bb28-d497024203a1,
            d9fbef25-ce0e-4675-91ea-4b1e969a80c2,
            12303d82-e59d-4f5e-bdfe-20e8c5f8b758,
            738ed341-b3c8-4f60-b1dc-daee44adcb96,
            283feb0d-9db9-404e-bb4e-139377502d25,
            93a839ea-0f69-4e36-adcc-869439b22070,
            7f139395-5117-4937-80b2-3e3ab6c8aa11,
            bf765fbc-c41c-4dd5-b796-51dd1ebd0d6b,
            a112b649-e104-4aab-930b-5ef186eeed18,
            8ab5df2e-0201-47de-b342-c013cc14b0d2,
            0ef3bab8-e78f-4527-a0b6-2a1fa192a7d8,
            1717528c-4b1e-4e9e-8775-47def673a192,
            12de6b21-7376-4d5b-9fa3-1c95a693ccbc,
            41372449-d667-4567-97a3-7fe769ddb82f,
            74b89914-f80e-41f5-918c-8961767caff4,
            c50e1d12-a125-46a2-b6bb-a88972a017b7,
            0ba51d36-57d7-4a6a-b3fb-a3127c2fb3da,
            609fc31e-2e54-480e-aa7f-313c8eee351b,
            f0248886-ebad-4dcc-bc22-53e3b3ef6a92,
            e54aa634-0939-43b7-b0ec-9486d5d2c12d,
            da15b4bd-1182-41a0-8ff1-c6ca3c04508f,
            998dc903-8605-4be0-bc75-46cc4438b1f2,
            32157422-7b06-4299-9e04-0668c2b18b7b,
            f29944ef-4b60-4145-a848-27d7fb879e1c,
            02cad4e4-00fc-4f8e-a3fb-9d9981f25ae1,
            ff737a85-24c6-4abf-a656-c2c01a4ebe70,
        ],
        ranks: {
            Councillor: [
                "Fruitloopins",
                "tuzzzie",
            ],
            Builder: [],
            Recruiter: [],
            Police: [],
            TaxExempt: [
                "the_kid",
            ],
            Treasurer: [],
            Realtor: [],
            Settler: [],
        },
    },
]
